src/btree/bt_cursor.c

```diff
@@ -932,37 +1003,14 @@ __wt_btcur_search_near(WT_CURSOR_BTREE *cbt, int *exactp)
         WT_ERR(__btcur_bounds_search_near_reposition(session, cbt));
 
     /*
-     * If we have a row-store page pinned, search it; if we don't have a page pinned, or the search
-     * of the pinned page doesn't find an exact match, search from the root. Unlike
-     * WT_CURSOR.search, ignore pinned pages in the case of column-store, search-near isn't an
-     * interesting enough case for column-store to add the complexity needed to avoid the tree
-     * search.
+     * For row-store search the pinned page if there is one. Unlike WT_CURSOR.search, ignore pinned
+     * pages in the case of column-store, search-near isn't an interesting enough case for
+     * column-store to add the complexity needed to avoid the tree search.
      */
-    valid = false;
-    if (btree->type == BTREE_ROW && __cursor_page_pinned(cbt, true)) {
-        __wt_txn_cursor_op(session);
-
-        /*
-         * Set the "insert" flag for row-store search; we may intend to position the cursor at the
-         * the end of the tree, rather than match an existing record. (LSM requires this semantic.)
-         */
-        WT_ERR(__cursor_row_search(cbt, true, cbt->ref, &leaf_found));
+    if (btree->type == BTREE_ROW)
+        WT_ERR(__btcur_search_near_row_pinned_page(cbt, &valid));
```

Comment:
Thanks for the refactor here - it is a nice change.

Could you also pull the `btree->type` check into the new function and get rid of the `row` from the name? If I were adding column store support, I'd prefer to do it within the bounds of a single function call from here, so I think the type check belongs further down.